### ISSUE TITLE
Combined chart bar width inconsistency

### ISSUE THREAD

**Issue by kanevP** [danielgindi/Charts#2765]
Using combined chart view I'm trying to display semi transparent bar data with no spacing between the bars, but I'm getting some unexpected behaviour: setting the `barWidth` of the set to 1.005 gets me chart with bars that start with spacing between them, towards the middle gets perfect and towards the end bars start to overlap and which results of a solid color - a combination of the 2 semi transparent.
[http://i.imgur.com/gh5i4cp.png](url) 

Running the same code on smaller screens gets me spacing and on bigger screen - overlap.

**Comment by liuxuan30** [comment]
cannot see the image. 
But I wonder if `barWidth` can exceeds 1?
> trying to display semi transparent bar data with no spacing between the bars

It confuses me if two bars don't have space between, they are one big bar instead... Then you just need one bar

**Comment by kanevP** [comment]
Fixed the image issue, I hope it clarifies what I'm trying to achieve.

In the image example the bars have 3 different values in the timeframe, but I can't have just 3 bars because the they should have different widths based on the time the data stayed like that and they should match the line data values time wise.

**Comment by liuxuan30** [comment]
>because the they should have different widths based on the time

all bars share the same width.

**Comment by kanevP** [comment]
Yes, reading back on my last comment maybe I was a bit too sleepy to post so let's get back to the original issue I'm experiencing: I'm trying to get bars with no spacing in between. If I set the `barWidth` of the data set to 1 I get the blue lines coming from the filled line data under the bars. If I set the `barWidth` to 1.1 I get light blue lines between the bars, coming from the overlap of the bars.
The image shows `barWidth` value of 1.005 and (I hope it's visible) on the left side we get space between the bars, which towards the middle is shrinking, right in the middle disappears completely (what I need) but towards the end the bars seem to start overlapping, causing the light blue line.

TLDR: `barWidth = 1.005` causes bars to overlap and have space in between in the same chart.

**Comment by liuxuan30** [comment]
I read several times and lost my mind then. Let's focus on `I'm trying to get bars with no spacing in between.`

How many bar data sets you have? Single bar chart, or multiple bars? (e.g. for each x value, how many bars you have)

**Comment by kanevP** [comment]
I have bar data with with one set of entries - single bar for each x value in a combined chart view.

**Comment by liuxuan30** [comment]
ok, then you have single bar chart. 

I tweak the ChartsDemo combine chart, it seems using bar width 1.05 will have no bar space between:
<img width="366" alt="image" src="https://user-images.githubusercontent.com/4375169/30357044-3e3d373e-986f-11e7-94b2-a4521323127d.png">

is this what are you asking? What I did is just tweak the code like:
```objective-c
    for (int index = 0; index < ITEM_COUNT; index++)
    {
        [entries1 addObject:[[BarChartDataEntry alloc] initWithX:index+0.5 y:(arc4random_uniform(25) + 25)]];
    }
    BarChartDataSet *set1 = [[BarChartDataSet alloc] initWithValues:entries1 label:@"Bar 1"];
    [set1 setColor:[UIColor colorWithRed:60/255.f green:220/255.f blue:78/255.f alpha:1.f]];
    set1.valueTextColor = [UIColor colorWithRed:60/255.f green:220/255.f blue:78/255.f alpha:1.f];
    set1.valueFont = [UIFont systemFontOfSize:10.f];
    set1.axisDependency = AxisDependencyLeft;
    float barWidth = 1.05f; // x2 dataset
    BarChartData *d = [[BarChartData alloc] initWithDataSets:@[set1]];
    d.barWidth = barWidth;
    return d;
```

**Comment by kanevP** [comment]
liuxuan30, you see no lines because the color you've set is solid, try

`[set1 setColor:[UIColor colorWithRed:60/255.f green:220/255.f blue:78/255.f alpha:.5f]];`

![screen shot 2017-09-13 at 10 20 26](https://user-images.githubusercontent.com/11072703/30364525-3fc1af38-986d-11e7-8de6-4c460bd875c7.png)

With `float barWidth = 1.05f;` you'll start seeing some solid green lines between the bars, which comes from the fact that `barWidth > 1` (overlap of two bars with color with 0.5 alpha).
If you do `barWidth = 1.0f` you'll  start seeing white lines (the chart view background color) between the bars. 

What I'm trying to achieve is having neither overlap of bars nor spacing between them.

I hope it's visible from the screenshot of my first post that in the beginning of the x axis I'm getting spacing between the bars (the dark lines with the color of what's under the bars), in the middle it looks good (no spacing, no overlap) and towards the end we start getting overlap of bars (the light lines).
![screen shot 2017-09-13 at 10 22 08](https://user-images.githubusercontent.com/11072703/30364561-724617fa-986d-11e7-9291-443d42e1257e.png)

**Comment by liuxuan30** [comment]
Yeah, I start understanding now.
I disabled grid lines and set bar width to 1.0, shows like:
<img width="350" alt="image" src="https://user-images.githubusercontent.com/4375169/30419102-c8f7bcd4-9967-11e7-801d-a1831118ff34.png">

If I look closely on the retina screen, I would see some very thin white lines (space) between SOME bars, that's to say it looks like a pixel alignment issue.

If we change to bar width 1.05, yes it have overlaps:
<img width="352" alt="image" src="https://user-images.githubusercontent.com/4375169/30419080-b67d9ef2-9967-11e7-94e0-b60cafa1e2c8.png">

It seems the bar width cannot exceed 1.0, as its comment:
```
    /// The width of the bars on the x-axis, in values (not pixels)
    ///
    /// **default**: 0.85
```
it's hard to explain, but it's like the total value is 100% but you set 105%, so it will overlap, so we should not use >1.0 value.

As for the spacing, need more time to investigate. You are more than welcome to look into pixel alignment issue. 

The point is to make sure the first bar rect's right x value equals next bar rect left x. 
in 
```swift
        prepareBuffer(dataSet: dataSet, index: index)
        trans.rectValuesToPixel(&_buffers[index].rects)
```
before apply matrix, the first two bar rects are:
```
(lldb) po _buffers[0].rects
▿ 12 elements
  ▿ 0 : (0.5, 46.0, 1.0, -46.0)
    ▿ origin : (0.5, 46.0)
      - x : 0.5
      - y : 46.0
    ▿ size : (1.0, -46.0)
      - width : 1.0
      - height : -46.0
  ▿ 1 : (1.5, 48.0, 1.0, -48.0)
    ▿ origin : (1.5, 48.0)
      - x : 1.5
      - y : 48.0
    ▿ size : (1.0, -48.0)
      - width : 1.0
      - height : -48.0
```
after converting to pixel system, the first two bar rects:
```
(lldb) po _buffers[0].rects
▿ 12 elements
  ▿ 0 : (40.936279296875, 319.767849959775, 26.64794921875, 178.497775040225)
    ▿ origin : (40.936279296875, 319.767849959775)
      - x : 40.936279296875
      - y : 319.767849959775
    ▿ size : (26.64794921875, 178.497775040225)
      - width : 26.64794921875
      - height : 178.497775040225
  ▿ 1 : (67.584228515625, 312.007077131939, 26.64794921875, 186.258547868061)
    ▿ origin : (67.584228515625, 312.007077131939)
      - x : 67.584228515625
      - y : 312.007077131939
    ▿ size : (26.64794921875, 186.258547868061)
      - width : 26.64794921875
      - height : 186.258547868061
```
you will see `40.936279296875 + 26.64794921875 = 67.584228515625` exactly:
```
(lldb) po _buffers[0].rects[0].origin.x + _buffers[0].rects[0].size.width - _buffers[0].rects[1].origin.x
0.0
```
So far it looks we did nothing wrong. 
I then added below test func:
```swift
        prepareBuffer(dataSet: dataSet, index: index)
        trans.rectValuesToPixel(&_buffers[index].rects)

        for i in 0..<_buffers.count
        {
            for j in 1..<_buffers[i].rects.count
            {
                let delta = _buffers[i].rects[j-1].origin.x + _buffers[i].rects[j-1].size.width - _buffers[i].rects[j].origin.x
                if delta != 0.0
                {
                    print("trap")
                }
            }
        }
```
to find out if any delta is not 0.0, however, it never print 'trap', you can try on your side, or point out if my code was wrong.

So it looks like it's not our fault for the space, we did what we could. The rest of it is CoreGraphics. You can try search or check the documentation if anything we could improve, or file a DTS request (each developer account has 2 chances) if they could offer better answer.

I tried a tricky way that is setting a small lineWidth like 0.01, to stroke the bar rects, that's to say:
```swift
            if drawBorder
            {
                context.setStrokeColor(borderColor.cgColor)
                context.setLineWidth(0.01)
                context.stroke(barRect)
            }
```
with this, it seems help somewhat, but the side effect is the line width is not small enough, you will see overlap again :( 0.01 is a fair value in ChartsDemo.

**Comment by liuxuan30** [comment]
So for you the quick around is change to solid color and use 1.01 bar width (pay attention to highlight as well), 
**or think another way**: use step line chart rather than bars. Our line chart offers stepped line (configured by `LinedataSet.mode`), seems a good replacement to me:
<img width="371" alt="image" src="https://user-images.githubusercontent.com/4375169/30421019-0847cd7e-996e-11e7-93a2-8e17afe34f4d.png">
